### PR TITLE
UI: do vipc recv in thread

### LINKED ISSUE #22257: C3: UI state updates at 10Hz
Camera view seems fine, but model and other state only updates at 10Hz

### PR DESCRIPTION
**Pull request by deanlee**
resolved #22257 #22427

### CODE REVIEW

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 1]
```diff
@@ -237,40 +220,55 @@ void CameraViewWidget::paintGL() {
...
-    buf = vipc_client->recv(nullptr, 0);
-    if (buf != nullptr) {
-      latest_frame = buf;
-      update();
-      emit frameUpdated();
+    VisionBuf *buf = vipc_client_->recv(nullptr, 1000);
+    if (buf) {
+      emit vipcThreadFrameReceived(buf);
+    } else {
+      LOGE("visionIPC receive timeout");
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:271]: let's remove this log. we already have the "slow frame" print in the onroad widget

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 2]
```diff
@@ -115,8 +113,7 @@ void CameraViewWidget::initializeGL() {
 
   program = new QOpenGLShaderProgram(context());
   bool ret = program->addShaderFromSourceCode(QOpenGLShader::Vertex, frame_vertex_shader);
-  assert(ret);
-  ret = program->addShaderFromSourceCode(QOpenGLShader::Fragment, frame_fragment_shader);
+  ret = ret && program->addShaderFromSourceCode(QOpenGLShader::Fragment, frame_fragment_shader);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:116]: think this was cleaner as is

**Comment by adeebshihadeh** [comment]
This looks a lot nicer!

**Comment by adeebshihadeh** [comment]
Still getting slow frames pretty consistently on this branch.

**Comment by deanlee** [comment]
> Still getting slow frames pretty consistently on this branch. Once that's fixed, looks good to merge.

It is caused by glTexImage2D in paintGL.  glTexImage2D takes 10-28ms to execute.
https://github.com/commaai/openpilot/blob/53e095343de56ff838f36ed9feda9f31ff6358b6/selfdrive/ui/qt/widgets/cameraview.cc#L225

**Comment by deanlee** [comment]
glTexImage2D will re-initialize the texture object. use glTexSubImage2D to speed things up a little.

**Comment by adeebshihadeh** [comment]
Ah, spent a bit of time trying to avoid the copy on C3 but seems like we're missing a bunch of GL extensions that would help here. Maybe you can take a stab at it too.

**Pull request merged**